pkg/agent/nodeportlocal/nplagent_test.go

```diff
@@ -0,0 +1,354 @@
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
+	npltest "github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/rules/testing"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/informers"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/tools/cache"
+)
+
+type nplAnnotation struct {
+	Podport  string
+	Nodeip   string
+	Nodeport string
+}
+
+func NewPortTable(c *gomock.Controller) *portcache.PortTable {
+	ptable := portcache.PortTable{StartPort: 40000, EndPort: 45000}
+	ptable.Table = make(map[int]portcache.NodePortData)
+
+	mockTable := npltest.NewMockPodPortRules(c)
+	mockTable.EXPECT().AddRule(gomock.Any(), gomock.Any()).AnyTimes()
+	mockTable.EXPECT().DeleteRule(gomock.Any(), gomock.Any()).AnyTimes()
+
+	ptable.PodPortRules = mockTable
+	return &ptable
+}
+
+const (
+	defaultPodName  = "test-pod"
+	defaultNS       = "default"
+	defaultNodeName = "test-node"
+	defaultHostIP   = "10.10.10.10"
+	defaultPodIP    = "192.168.32.10"
+	defaultPort     = 80
+)
+
+var kubeClient *k8sfake.Clientset
+var c *k8s.Controller
+var portTable *portcache.PortTable
+
+func getTestPod() corev1.Pod {
+	testPod := corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      defaultPodName,
+			Namespace: defaultNS,
+		},
+		Spec: corev1.PodSpec{
+			NodeName: defaultNodeName,
+			Containers: []corev1.Container{
+				{
+					Ports: []corev1.ContainerPort{
+						{
+							ContainerPort: int32(defaultPort),
+						},
+					},
+				},
+			},
+		},
+		Status: corev1.PodStatus{
+			HostIP: defaultHostIP,
+			PodIP:  defaultPodIP,
+		},
+	}
+	return testPod
+}
+
+// compareWithRetry : call fn() every second and compare returned value with expected value until timeout
+// Implements a minimal functionality similar to Eventually() of github.com/onsi/gomega
+func compareWithRetry(t *testing.T, fn interface{}, expected interface{}, timeout time.Duration) {
+	var actualValue []reflect.Value
+	var receivedData reflect.Value
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	timeCh := make(chan bool)
+	go func() {
+		time.Sleep(timeout)
+		timeCh <- true
+	}()
+	for {
+		select {
+		case <-timeCh:
+			if len(actualValue) > 0 {
+				receivedData = actualValue[0]
+			}
+			t.Fatalf("Timed out, actual value: %v did not match expected value: %v", receivedData, expected)
+		case _ = <-ticker.C:
+			actualValue = reflect.ValueOf(fn).Call([]reflect.Value{})
+			if len(actualValue) > 0 && reflect.DeepEqual(actualValue[0].Interface(), expected) {
+				return
+			}
+		}
+	}
+}
+
+func TestMain(t *testing.T) {
+	os.Setenv("NODE_NAME", defaultNodeName)
+	kubeClient = k8sfake.NewSimpleClientset()
+
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, 5*time.Second)
+	mockCtrl := gomock.NewController(t)
+	portTable = NewPortTable(mockCtrl)
+	c = k8s.NewNPLController(kubeClient, portTable)
+	c.OldObjStore = cache.NewStore(cache.MetaNamespaceKeyFunc)
```

Comment:
why is this needed? Isn't this the initialization that `NewNPLController` already does?